### DISCUSSION TITLE
Extract our Kubernetes middleware into a library or proxy?

### DISCUSSION THREAD

**Discussion by mattmoyer** [vmware-tanzu/pinniped#386] [Ideas]
We now have some neat middleware that allows multiple Pinniped instances to be deployed on the same cluster. It works by duplicating the `pinniped.dev` API groups under a custom suffix like `pinniped.example.com`, which means that you can have two copies of Pinniped that are effectively completely separate apps from a Kubernetes perspective.

The code we have to do this is fine for Pinniped, but maybe other folks that are writing Kubernetes controllers would also use it?

Two ideas:
- Extract it into a Go library (relatively easy).
- Extract it into an out-of-process API proxy (harder, but maybe more useful).

This discussion is a place to collect thoughts if anyone is interested in pursuing this further.

**Comment by cfryanr** [comment]
Something to keep in mind in case we would like to consider extracting this as a separate library... due to schedule constraints we leaned on our existing full suite of integration tests to give us confidence that this new middleware layer works for all of our use cases, and we didn't provide as much unit test coverage as we would normally provide. To make this a separate library it would probably need  more unit tests, and also need its own separate integration tests. Without that work, it would be hard to be able to evaluate and accept community PRs to the separate library with confidence. This would probably be very roughly 2-3 weeks of additional work for someone.
